src/Linters/UseCanonicalDataProviders.hack

```diff
@@ -0,0 +1,191 @@
+/*
+ *  Copyright (c) 2017-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace Facebook\HHAST;
+
+use namespace HH\Lib\{C, Dict, Str, Vec};
+
+final class UseCanonicalDataProvidersLinter extends AutoFixingASTLinter {
+  const type TContext = ClassishDeclaration;
+  const type TNode = FunctionDeclarationHeader;
+  // dict<provider, vec<(test, provider again)>>
+  const type TProviders = dict<string, vec<(string, string)>>;
+
+  const string T_DATAPROVIDER = 'Facebook\\HackTest\\DataProvider';
+
+  <<__Override>>
+  public function getLintErrorForNode(
+    this::TContext $context,
+    this::TNode $node,
+  ): ?ASTLintError {
+    $info = $this->getInfoAboutTestFile($context);
+    if ($info is null) {
+      return null;
+    }
+
+    list($data_providers, $hhast_methods) = tuple(
+      $info['data_providers'],
+      $info['hhast_methods'],
+    );
+
+    $method_name = $node->getName()->getText();
+
+    if (!C\contains_key($data_providers, $method_name)) {
+      return null;
+    }
+
+    $hhast_method = $hhast_methods[$method_name];
+
+    list($err, $_tuple_type) = $this->jugdeReturnTypes($hhast_method, $node);
+
+    if ($err is nonnull) {
+      return $err;
+    }
+
+    // TODO: Write validation for assignability from dataprovider to test method here.
```

Comment:
An exact match, ignoring trivia would probably be appropriate - IMO it's fine to say that if the dataprovider gives a MySubclass, the test must take a MySubclass, not a MySuperclass.

Reply:
Okay, just making sure, it is okay that this is a (false) positive / lint error:
Provider: `vec<(vec<string>, string)>`
Test: `function testThing(Traversable<string> $things, string $expected)`

Reply:
IMO this is completely fine for a linter

Reply:
one potential exception: perhaps the test function should be permitted to take generic parameters, in which case the types for that parameter are not checked.